relay/retry.go

```diff
@@ -0,0 +1,76 @@
+package relay
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+)
+
+// Buffers and retries operations, if the buffer is full operations are dropped.
+// Only tries one operation at a time, the next operation is not attempted
+// until success or timeout of the previous operation.
+// There is no delay between attempts of different operations.
+type retryBuffer struct {
+	buf chan retryOperation
+	max time.Duration
+
+	wg sync.WaitGroup
+}
+
+func newRetryBuffer(size int, max time.Duration) *retryBuffer {
+	r := &retryBuffer{
+		buf: make(chan retryOperation, size),
+		max: max,
+	}
+	r.wg.Add(1)
+	go r.run()
+	return r
+}
+
+// Stops the retryBuffer.
+// Subsequent calls to Rety will panic.
+// Actions currently in the buffer will still be tried.
+// Blocks until buffer is empty.
+func (r *retryBuffer) Stop() {
+	close(r.buf)
+	r.wg.Wait()
+}
+
+type retryOperation struct {
+	op   backoff.Operation
+	errC chan error
+}
+
+// Retry an operation, it is expected that one attempt has already
+// been made as this sleeps before trying again.
+func (r *retryBuffer) Retry(op backoff.Operation) error {
+	retry := retryOperation{
+		op:   op,
+		errC: make(chan error),
+	}
+
+	select {
+	case r.buf <- retry:
+		return <-retry.errC
+	default:
+		return errors.New("buffer full cannot retry")
+	}
+}
+
+func (r *retryBuffer) run() {
+	defer r.wg.Done()
+	for retry := range r.buf {
+		b := backoff.NewExponentialBackOff()
+		if r.max != 0 {
+			b.MaxElapsedTime = r.max
+		}
+		//We already tried once so sleep the initial interval
+		time.Sleep(b.InitialInterval)
+
+		// Possible to use RetryNotify to get a call back before each retry
+		// for counting failure stats etc
+		retry.errC <- backoff.Retry(retry.op, b)
```

Comment:
Use simple exponential backoff multiplier of 2 and initial internval 500ms make max interval configurable